{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from General.args import *\n",
    "from General.utils import *\n",
    "from PD_var_ILP.run_var_ilp import *\n",
    "from PD_mul_ILP.run_mul_ilp import *\n",
    "from PD_mul_greedy.run_mul_greedy import *\n",
    "from PD_single_LPath.run_pd_single_LPath import *\n",
    "import sys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set up gurobi optimization library\n",
    "\n",
    "\n",
    "First, create a file called gurobi.json with the details of your gurobi license in the main directory. The file should be in the following format:\n",
    "\n",
    "```json\n",
    "{\n",
    "  \"WLSACCESSID\": \"XXXXX\",\n",
    "  \"WLSSECRET\": \"XXXXX\",\n",
    "  \"LICENSEID\": 12345\n",
    "}\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Global program cmd arguments\n",
    "\n",
    "Set arguments to their default values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.argv = [\n",
    "        sys.argv[0],\n",
    "        \"--file_path\", \"test_inputs\",\n",
    "        \"--output\", \"Test_outputs\",\n",
    "        ]\n",
    "\n",
    "args = get_args()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define global variables for upstream and downstream regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Define global upstream and downstream regions\n",
    "UPSTREAM_NT = 'GCTAGTGGTGCTAGCCCCGCGAAATTAATACGACTCACTATAGGGTCTAGAAATAATTTTGTTTAACTTTAAGAAGGAGATATACAT'\n",
    "DOWNSTREAM_NT = 'GGAGGGTCTGGGGGAGGAGGCAGTGGCATGGTGAGCAAGGGCGAGGAGCTGTTCACCGGGGTGGTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAGGGCGAGGGCGATGCCACCTACGGCAAGCTGACCCTGAAGTTCATCTGCACCACCGGCAAGCTGCCCG'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run PD-Single-LPath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Define sequence\n",
    "First, we will define our mutreg region sequence, we will take the first 100 nucleotides of the coding sequence of the CXDAR human protein gene as a small test dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_name = 'CXDAR'\n",
    "mutreg_nt = 'ATGGCGCTCCTGCTGTGCTTCGTGCTCCTGTGCGGAGTAGTGGATTTCGCCAGAAGTTTGAGTATCACTACTCCTGAAGAGATGATTGAAAAAGCCAAAG'\n",
    "## Add upstream and downstream sequences\n",
    "full_sequence = UPSTREAM_NT + mutreg_nt + DOWNSTREAM_NT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run longest path algorithm\n",
    "\n",
    "We will runs PD-Single_LPath which runs a longest path algorithms on the primer graph, which is a DAG, to find the most efficient primers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating primer df\n",
      "Creating graph\n",
      "Saved CSV summary to: Test_outputs/PD_single_LPath_results.csv\n",
      "Saved path details to: Test_outputs/PD_single_LPath_selected_primers.json\n",
      "Results DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_name</th>\n",
       "      <th>graph_nodes</th>\n",
       "      <th>graph_edges</th>\n",
       "      <th>graph_time_sec</th>\n",
       "      <th>longest_path_efficiency</th>\n",
       "      <th>total_time_sec</th>\n",
       "      <th>num_primers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CXDAR</td>\n",
       "      <td>2550</td>\n",
       "      <td>160566</td>\n",
       "      <td>1.835</td>\n",
       "      <td>1.511491</td>\n",
       "      <td>6.713</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein_name  graph_nodes  graph_edges  graph_time_sec  \\\n",
       "0        CXDAR         2550       160566           1.835   \n",
       "\n",
       "   longest_path_efficiency  total_time_sec  num_primers  \n",
       "0                 1.511491           6.713            4  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_df, primer_paths = run_longest_path(full_sequence, mutreg_nt, protein_name,args) \n",
    "\n",
    "print(\"Results DataFrame:\")\n",
    "run_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Selected primer path and primer efficiency\n",
    "\n",
    "Print all selected primers in longest path and their efficiency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Primer Paths:\n",
      "(-80, -60, 'f')\n",
      "(96, 115, 'r')\n",
      "(70, 96, 'f')\n",
      "(244, 265, 'r')\n",
      "Primer effiency\n",
      "1.5114909044453393\n"
     ]
    }
   ],
   "source": [
    "print(\"Primer Paths:\")\n",
    "for path in primer_paths:\n",
    "    print(path)\n",
    "\n",
    "print(\"Primer effiency\")\n",
    "print(run_df['longest_path_efficiency'].prod())   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run PD-var-ILP\n",
    "\n",
    "We will run PD-var-ILP on multiple variants of the same protein.\n",
    "\n",
    "For this part, we will use the same sequence of the CXDAR protein as PD-single-LPath.\n",
    "We will use the default number of variants, which is 3, and the deafult maximum overlap length, which is 6. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating primer df\n",
      "Creating graph\n",
      "Running greedy algorithm\n",
      "Running ILP\n",
      "Set parameter WLSAccessID\n",
      "Set parameter WLSSecret\n",
      "Set parameter LicenseID to value 2499075\n",
      "Academic license 2499075 - for non-commercial use only - registered to jo___@live.biu.ac.il\n",
      "Number of Constraints: 5520\n",
      "Average Vars Per Constraint 80.94384057971014\n",
      "Graph edges:  160566\n",
      "Finished ILP Variable Creations\n",
      "0\n",
      "3\n",
      "7\n",
      "11\n",
      "15\n",
      "19\n",
      "23\n",
      "27\n",
      "31\n",
      "35\n"
     ]
    }
   ],
   "source": [
    "results_df, ilp_path, greedy_path  = run_var_ilp(full_sequence, mutreg_nt, protein_name,args)\n",
    "print(\"Variable ILP Results DataFrame:\")\n",
    "results_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compare PD-Var-ILP with greedy baseline\n",
    "\n",
    "We will compare the primers found by the ILP and by the greedy approach and their total efficiencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ILP chosen primers:\n",
      "Variant 1\n",
      "[\"(-86, -68, 'f')\", \"(87, 109, 'r')\", \"(59, 85, 'f')\", \"(244, 263, 'r')\"]\n",
      "Variant 2\n",
      "[\"(-74, -54, 'f')\", \"(106, 124, 'r')\"]\n",
      "Variant 3\n",
      "[\"(-58, -34, 'f')\", \"(123, 141, 'r')\"]\n",
      "ILP primer efficiency:  2.4237929216219447\n",
      "Greedy chosen primers:\n",
      "Variant 1\n",
      "[(-80, -60, 'f'), (96, 115, 'r'), (70, 96, 'f'), (244, 265, 'r')]\n",
      "Variant 2\n",
      "[(-58, -34, 'f'), (123, 141, 'r')]\n",
      "Variant 3\n",
      "[(-39, -9, 'f'), (136, 156, 'r')]\n",
      "Greedy primer efficiency:  1.514172485959763\n"
     ]
    }
   ],
   "source": [
    "print(\"ILP chosen primers:\")\n",
    "\n",
    "for i,path in enumerate(ilp_path):\n",
    "    print(\"Variant\", i+1)\n",
    "    print(path)\n",
    "\n",
    "print(\"ILP primer efficiency: \",results_df['ILP_primer_efficiency'].prod())\n",
    "\n",
    "\n",
    "print(\"Greedy chosen primers:\")\n",
    "for i,path in enumerate(greedy_path):\n",
    "    print(\"Variant\", i+1)\n",
    "    print(path)\n",
    "\n",
    "print(\"Greedy primer efficiency: \",results_df['greedy_primer_efficiency'].prod())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running PD-mul-ILP\n",
    "\n",
    "We will run PD-mul-ILP on multiple non-homologous proteins. For this part, we will add another partial 100-nt sequence of the SHP2 protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_name1 = 'CXDAR'\n",
    "mutreg_nt1 = 'ATGGCGCTCCTGCTGTGCTTCGTGCTCCTGTGCGGAGTAGTGGATTTCGCCAGAAGTTTGAGTATCACTACTCCTGAAGAGATGATTGAAAAAGCCAAAG'\n",
    "## Add upstream and downstream sequences\n",
    "full_sequence1 = UPSTREAM_NT + mutreg_nt1 + DOWNSTREAM_NT\n",
    "\n",
    "protein_name2 = 'SHP2'\n",
    "mutreg_nt2 = 'ATGACATCGCGGAGATGGTTTCACCCAAATATCACTGGTGTGGAGGCAGAAAACCTACTGTTGACAAGAGGAGTTGATGGCAGTTTTTTGGCAAGGCCTA'\n",
    "full_sequence2 = UPSTREAM_NT + mutreg_nt2 + DOWNSTREAM_NT\n",
    "\n",
    "## create protein dataset\n",
    "mutreg_regions = [mutreg_nt1, mutreg_nt2]\n",
    "sequences_nt = [full_sequence1, full_sequence2]\n",
    "protein_names = [protein_name1, protein_name2]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run PD-mul-ILP 2 non-homolous proteins parts\n",
    "PD-mul-ILP uses an integer linear program with forbidden pair constraints on all cross-hybridizing pairs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[STEP] Creating graphs for 2 proteins...\n",
      "Creating Graph for protein:  0 Protein name:  CXDAR\n",
      "Creating primer df\n",
      "Creating graph\n",
      "Creating Graph for protein:  1 Protein name:  SHP2\n",
      "Creating primer df\n",
      "Creating graph\n",
      "[DONE] Graphs created in 20.00 sec (peak 136.5 MB).\n",
      "[STEP] Finding forbidden pairs across proteins...\n",
      "Finding intra-protein forbidden pairs constraints\n",
      "[1/2] Processing CXDAR...\n",
      "[2/2] Processing SHP2...\n",
      "Number of intra-protein forbidden pairs constraints:  0\n",
      "Finding inter-protein forbidden pairs for 1 combinations...\n",
      "[1/1] Fractions of pairs considered\n",
      "Number of inter-protein forbidden pairs constraints:  0\n",
      "[DONE] Forbidden pairs in 24.85 sec (intra=0, inter=0).\n",
      "[STEP] Running ILP model...\n",
      "Creating ILP model...\n",
      "Set parameter WLSAccessID\n",
      "Set parameter WLSSecret\n",
      "Set parameter LicenseID to value 2499075\n",
      "Academic license 2499075 - for non-commercial use only - registered to jo___@live.biu.ac.il\n",
      "Adding constraints to ILP model...\n",
      "Gurobi Optimizer version 12.0.3 build v12.0.3rc0 (linux64 - \"Rocky Linux 9.6 (Blue Onyx)\")\n",
      "\n",
      "CPU model: Intel(R) Xeon(R) Gold 6338 CPU @ 2.00GHz, instruction set [SSE2|AVX|AVX2|AVX512]\n",
      "Thread count: 64 physical cores, 128 logical processors, using up to 32 threads\n",
      "\n",
      "Academic license 2499075 - for non-commercial use only - registered to jo___@live.biu.ac.il\n",
      "Optimize a model with 5100 rows, 321132 columns and 642264 nonzeros\n",
      "Model fingerprint: 0xd0d16cd9\n",
      "Variable types: 0 continuous, 321132 integer (321132 binary)\n",
      "Coefficient statistics:\n",
      "  Matrix range     [1e+00, 1e+00]\n",
      "  Objective range  [2e-04, 1e+00]\n",
      "  Bounds range     [1e+00, 1e+00]\n",
      "  RHS range        [1e+00, 1e+00]\n",
      "Found heuristic solution: objective -2.7100475\n",
      "Presolve removed 4212 rows and 308202 columns\n",
      "Presolve time: 0.88s\n",
      "Presolved: 888 rows, 12930 columns, 25860 nonzeros\n",
      "Found heuristic solution: objective 1.8129972\n",
      "Variable types: 0 continuous, 12930 integer (12930 binary)\n",
      "Performing another presolve...\n",
      "Presolve removed 888 rows and 12930 columns\n",
      "Presolve time: 0.02s\n",
      "Found heuristic solution: objective 3.3758909\n",
      "\n",
      "Explored 1 nodes (0 simplex iterations) in 1.01 seconds (2.02 work units)\n",
      "Thread count was 32 (of 128 available processors)\n",
      "\n",
      "Solution count 3: 3.37589 1.813 -2.71005 \n",
      "\n",
      "Optimal solution found (tolerance 1.00e-04)\n",
      "Best objective 3.375890943249e+00, best bound 3.375890943249e+00, gap 0.0000%\n",
      "Protein #CXDAR (4)\n",
      "[\"(-80, -60, 'f')\", \"(96, 115, 'r')\", \"(70, 96, 'f')\", \"(244, 265, 'r')\"]\n",
      "\n",
      "Protein #SHP2 (4)\n",
      "[\"(-80, -60, 'f')\", \"(98, 118, 'r')\", \"(69, 92, 'f')\", \"(244, 264, 'r')\"]\n",
      "\n",
      "[DONE] ILP status=2, objective=3.38, time=1.37 sec.\n",
      "[SAVE] Appended summary → Test_outputs/mul_ilp_results.csv\n",
      "[SAVE] Paths saved → Test_outputs/primers_N2.json\n",
      "PD-mul_ILP Results DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_proteins</th>\n",
       "      <th>graph_time_sec</th>\n",
       "      <th>ilp_num_vars</th>\n",
       "      <th>ilp_num_constraints</th>\n",
       "      <th>ilp_intra_forbidden_cnt</th>\n",
       "      <th>ilp_inter_forbidden_cnt</th>\n",
       "      <th>forbidden_time_sec</th>\n",
       "      <th>ilp_setup_time_sec</th>\n",
       "      <th>ilp_optimize_time_sec</th>\n",
       "      <th>ilp_feasibility</th>\n",
       "      <th>total_primer_efficiency</th>\n",
       "      <th>num_primers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "      <td>321132</td>\n",
       "      <td>5100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24.847219</td>\n",
       "      <td>13.202425</td>\n",
       "      <td>1.366277</td>\n",
       "      <td>FEASIBLE</td>\n",
       "      <td>3.375891</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   num_proteins  graph_time_sec  ilp_num_vars  ilp_num_constraints  \\\n",
       "0             2              20        321132                 5100   \n",
       "\n",
       "   ilp_intra_forbidden_cnt  ilp_inter_forbidden_cnt  forbidden_time_sec  \\\n",
       "0                        0                        0           24.847219   \n",
       "\n",
       "   ilp_setup_time_sec  ilp_optimize_time_sec ilp_feasibility  \\\n",
       "0           13.202425               1.366277        FEASIBLE   \n",
       "\n",
       "   total_primer_efficiency  num_primers  \n",
       "0                 3.375891            8  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "run_df, primer_paths = run_mul_ilp(mutreg_regions, sequences_nt, protein_names, args)\n",
    "\n",
    "print(\"PD-mul_ILP Results DataFrame:\")\n",
    "run_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Primer slection and total efficiency\n",
    "\n",
    "We will look at the primer paths found for each protein by PD-mul-ILP, and the total efficiency of the selected primers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chosen primers:\n",
      "Protein: CXDAR\n",
      "[\"(-80, -60, 'f')\", \"(96, 115, 'r')\", \"(70, 96, 'f')\", \"(244, 265, 'r')\"]\n",
      "Protein: SHP2\n",
      "[\"(-80, -60, 'f')\", \"(98, 118, 'r')\", \"(69, 92, 'f')\", \"(244, 264, 'r')\"]\n",
      "Total primer efficiency:\n",
      "3.3758909432490496\n"
     ]
    }
   ],
   "source": [
    "print(\"Chosen primers:\")\n",
    "\n",
    "for protein, path in primer_paths.items():\n",
    "    print(\"Protein:\", protein)\n",
    "    print(path)\n",
    "\n",
    "total_efficiency = run_df['total_primer_efficiency'].prod()   \n",
    "print(\"Total primer efficiency:\")\n",
    "print(total_efficiency) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running PD-mul-Greedy\n",
    "\n",
    "We will run PD-mul-Greedy, which uses a greedy algorithm to find efficient primers paths for each protein, on multiple non-homologous proteins. \n",
    "We will use the same proteins as we used for the PD-mul-ILP run. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO] Processing protein 0/2: CXDAR\n",
      "Creating primer df\n",
      "Creating graph\n",
      "Creating primer df\n",
      "Creating graph\n",
      " Saved summary to: Test_outputs/summary.csv\n",
      " Saved per-protein metrics to: Test_outputs/per_protein_metrics.csv\n",
      " Saved paths to: Test_outputs/primers_per_protein.json\n",
      "PD-mul_greedy Results DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_proteins</th>\n",
       "      <th>total_primer_efficiency</th>\n",
       "      <th>greedy_time_sec</th>\n",
       "      <th>cross_hybridizations_cnt</th>\n",
       "      <th>proteins_with_reiterations_cnt</th>\n",
       "      <th>total_reiterations</th>\n",
       "      <th>unresolved_proteins_cnt</th>\n",
       "      <th>unresolved_proteins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>3.375891</td>\n",
       "      <td>13.527</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   num_proteins  total_primer_efficiency  greedy_time_sec  \\\n",
       "0             2                 3.375891           13.527   \n",
       "\n",
       "   cross_hybridizations_cnt  proteins_with_reiterations_cnt  \\\n",
       "0                         0                               0   \n",
       "\n",
       "   total_reiterations  unresolved_proteins_cnt unresolved_proteins  \n",
       "0                   0                        0                      "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "run_df, paths = run_mul_greedy(sequences_nt,mutreg_regions, protein_names, args)\n",
    "\n",
    "print(\"PD-mul_greedy Results DataFrame:\")\n",
    "run_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Primers found for each protein total efficiency\n",
    "\n",
    "We will examine the total efficiency and the primers found for each of the proteins by the greedy algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein: CXDAR\n",
      "[(-80, -60, 'f'), (96, 115, 'r'), (70, 96, 'f'), (244, 265, 'r')]\n",
      "Protein: SHP2\n",
      "[(-80, -60, 'f'), (98, 118, 'r'), (69, 92, 'f'), (244, 264, 'r')]\n",
      "Total primer efficiency:\n",
      "3.3758909432490496\n"
     ]
    }
   ],
   "source": [
    "for protein, path in paths.items():\n",
    "    print(\"Protein:\", protein)\n",
    "    print(path)\n",
    "\n",
    "total_efficiency = run_df['total_primer_efficiency'].prod()   \n",
    "print(\"Total primer efficiency:\")\n",
    "print(total_efficiency)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "primer_env3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.25"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
