{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from General.args import *\n",
    "from General.utils import *\n",
    "from PD_var_ILP.run_var_ilp import *\n",
    "from PD_mul_ILP.run_mul_ilp import *\n",
    "from PD_mul_greedy.run_mul_greedy import *\n",
    "from PD_single_LPath.run_pd_single_LPath import *\n",
    "import sys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set up gurobi optimization library\n",
    "\n",
    "\n",
    "First, create a file called gurobi.json with the details of your gurobi license in the main directory. The file should be in the following format:\n",
    "\n",
    "```json\n",
    "{\n",
    "  \"WLSACCESSID\": \"XXXXX\",\n",
    "  \"WLSSECRET\": \"XXXXX\",\n",
    "  \"LICENSEID\": 12345\n",
    "}\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Global program cmd arguments\n",
    "\n",
    "Set arguments to their default values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.argv = [\n",
    "        sys.argv[0],\n",
    "        \"--file_path\", \"test_inputs\",\n",
    "        \"--output\", \"Test_outputs\",\n",
    "        ]\n",
    "\n",
    "args = get_args()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define global variables for upstream and downstream regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Define global upstream and downstream regions\n",
    "UPSTREAM_NT = 'GCTAGTGGTGCTAGCCCCGCGAAATTAATACGACTCACTATAGGGTCTAGAAATAATTTTGTTTAACTTTAAGAAGGAGATATACAT'\n",
    "DOWNSTREAM_NT = 'GGAGGGTCTGGGGGAGGAGGCAGTGGCATGGTGAGCAAGGGCGAGGAGCTGTTCACCGGGGTGGTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAGGGCGAGGGCGATGCCACCTACGGCAAGCTGACCCTGAAGTTCATCTGCACCACCGGCAAGCTGCCCG'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run PD-Single-LPath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Define sequence\n",
    "First, we will define our mutreg region sequence, we will take the first 100 nucleotides of the coding sequence of the CXDAR human protein gene as a small test dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_name = 'CXDAR'\n",
    "mutreg_nt = 'ATGGCGCTCCTGCTGTGCTTCGTGCTCCTGTGCGGAGTAGTGGATTTCGCCAGAAGTTTGAGTATCACTACTCCTGAAGAGATGATTGAAAAAGCCAAAG'\n",
    "## Add upstream and downstream sequences\n",
    "full_sequence = UPSTREAM_NT + mutreg_nt + DOWNSTREAM_NT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run longest path algorithm\n",
    "\n",
    "We will runs PD-Single_LPath which runs a longest path algorithms on the primer graph, which is a DAG, to find the most efficient primers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_df, primer_paths = run_longest_path(full_sequence, mutreg_nt, protein_name,args) \n",
    "\n",
    "print(\"Results DataFrame:\")\n",
    "run_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Selected primer path and primer efficiency\n",
    "\n",
    "Print all selected primers in longest path and their efficiency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Primer Paths:\")\n",
    "for path in primer_paths:\n",
    "    print(path)\n",
    "\n",
    "print(\"Primer effiency\")\n",
    "run_df['longest_path_efficiency']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run PD-var-ILP\n",
    "\n",
    "We will run PD-var-ILP on multiple variants of the same protein.\n",
    "\n",
    "For this part, we will use the same sequence of the CXDAR protein as PD-single-LPath.\n",
    "We will use the default number of variants, which is 3, and the deafult maximum overlap length, which is 6. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df, ilp_path, greedy_path  = run_var_ilp(full_sequence, mutreg_nt, protein_name,args)\n",
    "print(\"Variable ILP Results DataFrame:\")\n",
    "results_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compare PD-Var-ILP with greedy baseline\n",
    "\n",
    "We will compare the primers found by the ILP and by the greedy approach and their total efficiencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"ILP chosen primers:\")\n",
    "\n",
    "for i,path in enumerate(ilp_path):\n",
    "    print(\"Variant\", i+1)\n",
    "    print(path)\n",
    "\n",
    "print(\"ILP primer efficiency: \",results_df['ILP_primer_efficiency'])\n",
    "\n",
    "\n",
    "print(\"Greedy chosen primers:\")\n",
    "for i,path in enumerate(greedy_path):\n",
    "    print(\"Variant\", i+1)\n",
    "    print(path)\n",
    "\n",
    "print(\"Greedy primer efficiency: \",results_df['greedy_primer_efficiency'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running PD-mul-ILP\n",
    "\n",
    "We will run PD-mul-ILP on multiple non-homologous proteins. For this part, we will add another partial 100-nt sequence of the SHP2 protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_name1 = 'CXDAR'\n",
    "mutreg_nt1 = 'ATGGCGCTCCTGCTGTGCTTCGTGCTCCTGTGCGGAGTAGTGGATTTCGCCAGAAGTTTGAGTATCACTACTCCTGAAGAGATGATTGAAAAAGCCAAAG'\n",
    "## Add upstream and downstream sequences\n",
    "full_sequence1 = UPSTREAM_NT + mutreg_nt1 + DOWNSTREAM_NT\n",
    "\n",
    "protein_name2 = 'SHP2'\n",
    "mutreg_nt2 = 'ATGACATCGCGGAGATGGTTTCACCCAAATATCACTGGTGTGGAGGCAGAAAACCTACTGTTGACAAGAGGAGTTGATGGCAGTTTTTTGGCAAGGCCTA'\n",
    "full_sequence2 = UPSTREAM_NT + mutreg_nt2 + DOWNSTREAM_NT\n",
    "\n",
    "## create protein dataset\n",
    "mutreg_regions = [mutreg_nt1, mutreg_nt2]\n",
    "sequences_nt = [full_sequence1, full_sequence2]\n",
    "protein_names = [protein_name1, protein_name2]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run PD-mul-ILP 2 non-homolous proteins parts\n",
    "PD-mul-ILP uses an integer linear program with forbidden pair constraints on all cross-hybridizing pairs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "run_df, primer_paths = run_mul_ilp(mutreg_regions, sequences_nt, protein_names, args)\n",
    "\n",
    "print(\"PD-mul_ILP Results DataFrame:\")\n",
    "run_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Primer slection and total efficiency\n",
    "\n",
    "We will look at the primer paths found for each protein by PD-mul-ILP, and the total efficiency of the selected primers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Chosen primers:\")\n",
    "\n",
    "for protein, path in primer_paths.items():\n",
    "    print(\"Protein:\", protein)\n",
    "    print(path)\n",
    "\n",
    "total_efficiency = run_df['total_primer_efficiency'].prod()   \n",
    "print(\"Total primer efficiency:\")\n",
    "print(total_efficiency) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running PD-mul-Greedy\n",
    "\n",
    "We will run PD-mul-Greedy, which uses a greedy algorithm to find efficient primers paths for each protein, on multiple non-homologous proteins. \n",
    "We will use the same proteins as we used for the PD-mul-ILP run. We will use the same proteins as PD-mul-ILP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "run_df, paths = run_mul_greedy(sequences_nt,mutreg_regions, protein_names, args)\n",
    "\n",
    "print(\"PD-mul_greedy Results DataFrame:\")\n",
    "run_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Primers found for each protein total efficiency\n",
    "\n",
    "We will examine the total efficiency and the primers found for each of the proteins by the greedy algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for protein, path in paths.items():\n",
    "    print(\"Protein:\", protein)\n",
    "    print(path)\n",
    "\n",
    "total_efficiency = run_df['total_primer_efficiency'].prod()   \n",
    "print(\"Total primer efficiency:\")\n",
    "print(total_efficiency)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "primer_env2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.25"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
